ceci/minirunner.py

```diff
@@ -0,0 +1,297 @@
+
+    def __hash__(self):
+        return hash(self.id)
+
+    def assign(self):
+        """Set this node as assigned to a job"""
+        self.is_assigned = True
+
+    def free(self):
+        """Set this node as no longer assigned to a job"""
+        self.is_assigned = False
+
+
+class Job:
+    """Small wrapper for a job to be run by minirunner, incorporating
+    the command line and the resources needed to run it.
+    """
+    def __init__(self, name, cmd, nodes, cores):
+        """Create a node.
+
+        Parameters
+        ----------
+        name: str
+            Name for the job
+
+        cmd: str
+            Command line to execute for the job
+
+        cores: int
+            Number of cores needed for the job
+
+        nodes: int
+            Number of nodes needed for the job
+
+        Attributes
+        ----------
+
+        name: str
+            Name for the node
+
+        cmd: str
+            Command line to execute for the job
+
+        cores: int
+            Number of cores needed for the job
+
+        nodes: int
+            Number of nodes needed for the job
+        """        
+        self.name = name
+        self.cmd = cmd
+        self.nodes = nodes
+        self.cores = cores
+
+    def __str__(self):
+        return f"<Job {self.name}>"
+
+
+class Runner:
+    """The main pipeline runner class.
+
+    User is responsible for supplying the graph of jobs to run on it, etc.
+
+    Attributes
+    ----------
+
+    nodes: list[Node]
+        Nodes available to run on
+
+    job_graph: dict{Job:Job}
+        Indicates what jobs are to be run and which other jobs they depend on
+
+    completed_jobs: list[Job]
+        Jobs that have finished
+
+    running: list[Job]
+        Jobs currently running
+
+    queued_jobs: list[Job]
+        Jobs waiting to be run
+
+    log_dir: str
+        Dir where the logs are put
+
+    """
+    def __init__(self, nodes, job_graph, log_dir):
+        """Create a Runner
+
+        Parameters
+        ----------
+        nodes: list[Node]
+            Nodes available to run on
+
+        job_graph: dict{Job:Job}
+            Indicates what jobs are to be run and which other jobs they depend on
+
+        log_dir: str
+            Dir where the logs are put        
+
+        """
+        self.nodes = nodes
+        self.job_graph = job_graph
+        self.completed_jobs = []
+        self.running = []
+        self.log_dir = log_dir
+        self.queued_jobs = list(job_graph.keys())
+
+
+    def run(self, interval):
+        """Launch the pipeline.
+
+        Parameters
+        ----------
+        interval: int
+            How long to wait between each poll of the jobs.
+
+        """
+        status = WAITING
+        while status == WAITING:
+            status = self._update()
+            try:
+                time.sleep(interval)
+            except KeyboardInterrupt:
+                self.abort()
+                raise
```

Comment:
Thanks - I've combined the two `except`s into a single call.
Also I only want to `abort` in the event of an error - it should be clean already otherwise